### PR TITLE
[NFC] TypeCheckType No Longer Returns Null

### PR DESCRIPTION
**Pull request by CodaFi**
Make it structurally impossible for TypeCheckType to return the null type to both itself and its callers. A large amount of this is based on the work in #29780, but there's some additional deletions and structural assertions here.

Resolves [SR-12085](https://bugs.swift.org/browse/SR-12085) and rdar://58879303

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci test source compatibility

File: validation-test/compiler_crashers_2_fixed/0163-sr8033.swift [hunk 1]
```diff
@@ -7,4 +7,4 @@ protocol P1 {
 }
 extension Foo: P1 where A : P1 {} // expected-error {{unsupported recursion for reference to associated type 'A' of type 'Foo<T>'}}
 // expected-error@-1 {{type 'Foo<T>' does not conform to protocol 'P1'}}
-// expected-error@-2 {{type 'Foo<T>' in conformance requirement does not refer to a generic parameter or associated type}}
+// expected-error@-2 {{type '<<error type>>' in conformance requirement does not refer to a generic parameter or associated type}}
```
Comment (slavapestov) [inline validation-test/compiler_crashers_2_fixed/0163-sr8033.swift:10]: Why did this change?

**Review by slavapestov — APPROVED** [review]

File: validation-test/compiler_crashers_2_fixed/0163-sr8033.swift (hunk 1 above)
Comment (AnthonyLatsis) [reply validation-test/compiler_crashers_2_fixed/0163-sr8033.swift:10]: Looks like the result of a `!ty` -> `ty->hasError()` transition somewhere in `TypeCheckType.cpp`. The error message is bogus though, anyway..

Comment (CodaFi) [reply validation-test/compiler_crashers_2_fixed/0163-sr8033.swift:10]: The second time we come around to add these requirements, the base is invalidated somewhere else in the stack. I think we should just not emit this diagnostic if the subject is an `ErrorType` - it doesn't make sense here in the first place.

**Comment by CodaFi** [comment]
Lost connection to macOS nodes

@swift-ci smoke test macOS

**Comment by CodaFi** [comment]
Release builder died.

@swift-ci test source compatibility Release

**Comment by CodaFi** [comment]
@swift-ci smoke test macOS

**Pull request merged**